packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -14,71 +14,249 @@
+   *
+   * If no token is specified, anonymous access is used.
+   */
+  token?: string;
+
+  /**
+   * Username
+   */
+  username?: string;
+
+  /**
+   * Password
+   */
+  password?: string;
 };
 
-function readConfig(config: Config): Options[] {
-  const optionsArr = Array<Options>();
+export function getApiRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {};
+
+  if (provider.token) {
+    headers.Authorization = `Bearer ${provider.token}`;
+  } else if (provider.username && provider.password) {
+    headers.Authorization = `Basic ${Buffer.from(
+      `${provider.username}:${provider.password}`,
+      'utf8',
+    ).toString('base64')}`;
+  }
+
+  return {
+    headers,
+  };
+}
+
+export function getRawRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {};
+  if (provider.token) {
+    headers.Authorization = `Bearer ${provider.token}`;
+  } else if (provider.username && provider.password) {
+    headers.Authorization = `Basic ${Buffer.from(
+      `${provider.username}:${provider.password}`,
+      'utf8',
+    ).toString('base64')}`;
+  }
+
+  return {
+    headers,
+  };
+}
+
+// Converts for example
+// from: https://bitbucket.org/orgname/reponame/src/master/file.yaml
+// to:   https://api.bitbucket.org/2.0/repositories/orgname/reponame/src/master/file.yaml
+export function getApiUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+    if (
+      !owner ||
+      !name ||
+      (filepathtype !== 'browse' &&
+        filepathtype !== 'raw' &&
+        filepathtype !== 'src')
+    ) {
+      throw new Error('Invalid Bitbucket URL or file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+
+    if (provider.host === 'bitbucket.org') {
+      if (!ref) {
+        throw new Error('Invalid Bitbucket URL or file path');
+      }
+      return new URL(
+        `${provider.apiBaseUrl}/repositories/${owner}/${name}/src/${ref}/${pathWithoutSlash}`,
+      );
+    }
+    return new URL(
+      `${provider.apiBaseUrl}/projects/${owner}/repos/${name}/raw/${pathWithoutSlash}?at=${ref}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+// Converts for example
+// from: https://bitbucket.org/orgname/reponame/src/master/file.yaml
+// to:   https://api.bitbucket.org/2.0/repositories/orgname/reponame/src/master/file.yaml
+export function getRawUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+
+    if (
+      !owner ||
+      !name ||
+      (filepathtype !== 'browse' &&
+        filepathtype !== 'raw' &&
+        filepathtype !== 'src')
+    ) {
+      throw new Error('Invalid Bitbucket URL or file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+
+    if (provider.host === 'bitbucket.org') {
```

Comment:
Oh is the url pattern consistently different on public and private bitbucket? And not depending on what version you have on the private instance? Interesting

Reply:
Yes, the Bitbucket Server product only support the v1 api, while the cloud version supports v2 and only some parts of v1.
In addition, Atlassian have just [released their plans](https://www.atlassian.com/migration/journey-to-cloud) to retire their on-prem products so Bitbucket Server will be no more in a couple of years.

Reply:
I see, thanks for the insights